### PR TITLE
Regulator power state control

### PR DESCRIPTION
**Pull request by aasinclair**
Generic regulator power state interface, and associated support for the nPM1300

### CODE REVIEW

File: include/zephyr/drivers/regulator.h [hunk 1]
```diff
@@ -279,6 +287,38 @@ static inline int regulator_parent_dvs_state_set(const struct device *dev,
...
+ * @param state power state (vendor specific identifier).
+ * @param state_cfg Timeout configuration (vendor specific).
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If given state is not supported.
+ * @retval -EPERM If state can't be changed by software.
+ * @retval -ENOSYS If function is not implemented.
+ * @retval -errno In case of any other error.
+ */
+static inline int regulator_parent_power_state_set(const struct device *dev,
```
Comment (danieldegrasse) [inline include/zephyr/drivers/regulator.h:308]: I'm a bit confused what the intention of this API is. This looks like a feature where the entire regulator can be moved from a high to low power state (via a pin transition or via software). The [regulator_parent_dvs_state_set](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/regulator.h#L269) API was designed to cover the use case where an entire regulator needs to be reconfigured. Could this feature of the NPM1300 fit into that API?

**Review by gmarull — COMMENTED** [review]
quick comment here: since _ship_ mode seems to be a feature shared by multiple PMICs (PCA9420 in-tree has it for example), I think we can probably have an specific API for it or at least a generic property, same pattern as proposed for the new charger API.

I'm not sure DVS applies for ship/hibernate features, DVS (Dynamic Voltage Scaling) functionality is defined like this in the API:

```
 * Some PMICs feature DVS (Dynamic Voltage Scaling) by allowing to program the
 * voltage level for multiple states. Such states may be automatically changed
 * by hardware using GPIO pins. Certain MCUs even allow to automatically
 * configure specific output pins when entering low-power modes so that PMIC
 * state is changed without software intervention. This API can be used when
 * state needs to be changed by software.
```

**Comment by aasinclair** [comment]
Thinking about this in light of the refactoring in https://github.com/zephyrproject-rtos/zephyr/pull/58229, it would make more sense to have the power control for the entire device at the MFD level.
Would it make sense to add a power management action handler at that level to implement PM_DEVICE_ACTION_TURN_OFF instead?

**Comment by gmarull** [comment]
> Thinking about this in light of the refactoring in #58229, it would make more sense to have the power control for the entire device at the MFD level. Would it make sense to add a power management action handler at that level to implement PM_DEVICE_ACTION_TURN_OFF instead?

If PM needs to happen at a device level (or requires coordination between its multiple IPs), it sounds like a good approach. Note that `PM_DEVICE_ACTION_TURN_OFF` is only used in the context of power domains, devices are either active or suspended. Naming isn't great, we should likely reconsider replacing "TURN_OFF" with something else...

**Comment by aasinclair** [comment]
As the power state relates to the entire device, not just the regulators, this will be handled by the new MFD driver instead.

**Comment by aasinclair** [comment]
I've updated this to just include the ship mode command.
Hibernate is rather specific to the nPM1300, but ship mode is common across many PMIC devices.

**Review by gmarull — APPROVED** [review]

**Review by fabiobaltieri — COMMENTED** [review]
Cool! Wondering if the API should not have a parameter for the device specific shipping mode option, but on a second though it should probably be done separately and keep the API simple (I'm looking at https://www.ti.com/lit/ds/symlink/bq25180.pdf specifically)

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**